java-checks/src/main/java/org/sonar/java/checks/spring/MissingPathVariableAnnotationCheck.java

```diff
@@ -0,0 +1,374 @@
+      var values = method.symbol().metadata().valuesForAnnotation(fullyQualifiedName);
+      if (values == null || !MAPPING_ANNOTATIONS.contains(fullyQualifiedName)) {
+        continue;
+      }
+
+      @Nullable
+      var templatesVars = templateVariablesFromMapping(values);
+
+      if (templatesVars != null) {
+        templateVariables.add(new UriInfo<>(ann, templatesVars));
+      }
+    }
+
+    // we handle the case where a path variable doesn't match to an uri parameter (/{aParam}/)
+    Set<String> allTemplateVariables = templateVariables.stream()
+      .flatMap(uri -> uri.value().stream())
+      .collect(Collectors.toSet());
+    allTemplateVariables.addAll(requestMappingTemplateVars);
+
+    methodParameters.stream()
+      .filter(v -> !allTemplateVariables.contains(v.value()))
+      .filter(v -> !v.parameter().symbol().type().is(MAP))
+      .forEach(v -> reportIssue(v.parameter(), String.format("Bind method parameter \"%s\" to a template variable.", v.value())));
+
+    if (containsTypeMapAsParameter(method)) {
+      /*
+       * If any of the method parameters is a map, we assume all path variables are captured
+       * and there is no mismatch with path variables in the request mapping.
+       */
+      return;
+    }
+
+    Set<String> allPathVariables = methodParameters.stream()
+      .map(ParameterInfo::value)
+      .collect(Collectors.toSet());
+    allPathVariables.addAll(modelAttributeMethodParameters);
+
+    templateVariables.stream()
+      .filter(uri -> !allPathVariables.containsAll(uri.value()))
+      .forEach(uri -> {
+        Set<String> unbind = new HashSet<>(uri.value());
+        unbind.removeAll(allPathVariables);
+        reportIssue(
+          uri.request(),
+          "Bind template variable \"" + String.join("\", \"", unbind) + "\" to a method parameter.");
+      });
+
+  }
+
+  private static boolean containsTypeMapAsParameter(MethodTree method) {
+    return method.parameters().stream()
+      .filter(parameter -> parameter.symbol().metadata().isAnnotatedWith(PATH_VARIABLE_ANNOTATION))
+      .anyMatch(parameter -> {
+        Type type = parameter.type().symbolType();
+        return type.isSubtypeOf(MAP);
+      });
+  }
+
+  @Nullable
+  private static Set<String> templateVariablesFromMapping(List<SymbolMetadata.AnnotationValue> values) {
+    Map<String, Object> nameToValue = values.stream()
+      .collect(Collectors.toMap(SymbolMetadata.AnnotationValue::name, SymbolMetadata.AnnotationValue::value));
+    List<String> path = arrayOrString(nameToValue.get("path"));
+    List<String> value = arrayOrString(nameToValue.get("value"));
+
+    if (path != null || value != null) {
+      List<String> paths = path != null ? path : value;
+      return paths.stream()
+        .map(PathPatternParser::parsePathVariables)
+        .flatMap(Collection::stream)
+        .collect(Collectors.toSet());
+    } else {
+      return null;
```

Comment:
I think this method can return an empty set as a fallback, so that we don't have to handle the null case at line ~75.
We can also remove the `@Nullable` annotation from here then
```suggestion
      return new HashSet<>();
```

Reply:
nice catch :100: